src/script/zcash_script.cpp

```diff
@@ -232,6 +232,94 @@ unsigned int zcash_script_legacy_sigop_count(
     }
 }
 
+zcash_script_uint160 zcash_script_transparent_output_address_precomputed(
+    const void* pre_preTx,
+    unsigned int nOut,
+    zcash_script_type* type,
+    zcash_script_error* err)
+{
+    const PrecomputedTransaction* preTx = static_cast<const PrecomputedTransaction*>(pre_preTx);
+
+    zcash_script_uint160 r;
+    memset(&r, 0, sizeof(r));
+
+    if (nOut >= preTx->tx.vout.size()) {
+        set_error(err, zcash_script_ERR_TX_INDEX);
+        return r;
+    }
+
+    const CTxOut& out = preTx->tx.vout[nOut];
+    CScript::ScriptType scriptType = out.scriptPubKey.GetType();
+    switch (scriptType) {
+        case CScript::ScriptType::P2PKH:
+            if (type) {
+                *type = zcash_script_TYPE_P2PKH;
+            }
+            break;
+        case CScript::ScriptType::P2SH:
+            if (type) {
+                *type = zcash_script_TYPE_P2SH;
+            }
+            break;
+        default:
+            set_error(err, zcash_script_ERR_TX_INVALID_SCRIPT);
+            return r;
+    }
+    const uint160 addr = out.scriptPubKey.AddressHash();
+
+    assert(sizeof(r.value) == addr.size());
+    memcpy(r.value, addr.begin(), addr.size());
+    set_error(err, zcash_script_ERR_OK);
+    return r;
+}
+
+zcash_script_uint160 zcash_script_transparent_output_address(
+    const unsigned char *txTo,
+    unsigned int txToLen,
+    unsigned int nOut,
+    zcash_script_type* type,
+    zcash_script_error* err)
+{
+    TxInputStream stream(SER_NETWORK, PROTOCOL_VERSION, txTo, txToLen);
+    CTransaction tx;
+    zcash_script_uint160 r;
+    memset(&r, 0, sizeof(r));
+
+    stream >> tx;
+    if (nOut >= tx.vout.size()) {
+        set_error(err, zcash_script_ERR_TX_INDEX);
+        return r;
+    }
+    if (GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION) != txToLen) {
+        set_error(err, zcash_script_ERR_TX_SIZE_MISMATCH);
+        return r;
+    }
+
+    const CTxOut& out = tx.vout[nOut];
+    CScript::ScriptType scriptType = out.scriptPubKey.GetType();
+    switch (scriptType) {
+        case CScript::ScriptType::P2PKH:
+            if (type) {
+                *type = zcash_script_TYPE_P2PKH;
+            }
+            break;
+        case CScript::ScriptType::P2SH:
+            if (type) {
+                *type = zcash_script_TYPE_P2SH;
+            }
+            break;
+        default:
+            set_error(err, zcash_script_ERR_TX_INVALID_SCRIPT);
+            return r;
+    }
+    const uint160 addr = out.scriptPubKey.AddressHash();
```

Comment:
This PR (and the resulting FFI logic either side) seems like a lot of effort to go through just to figure out if the `scriptPubKey` of a transparent output is P2PKH or P2SH. All that the methods you're calling do is pattern-match on the script bytes, and this can trivially be replicated in Rust (unlike the script engine, exposing of which is the main purpose of `libzcash_script`). In fact, we have already done so in `zcash_primitives::legacy::Script::address`: https://github.com/zcash/librustzcash/blob/c423167b30f7b7afeeabe09ae5abceb49ce7d06e/zcash_primitives/src/legacy.rs#L45-L64

So I'm -1 on including this change in `libzcash_script` given its only current use case is the `zcash_script` Rust library (or rather, its `zebra` consumer), and instead suggest that `zebra` use its existing dependency on the `zcash_primitives` crate for this functionality. If there were a C consumer who needed this pattern-matching functionality then I might reconsider (because then you're already operating in a memory-unsafe domain and paying the costs that the FFI require).

Reply:
I kinda of reached a similar conclusion while looking at how it was implemented, but I didn't realize we could get the same logic from `zcash_primitives`. I'll look into using it, thanks for pointing this out!